src/NHibernate.Test/NHSpecificTest/NH2176/Fixture.cs

```diff
@@ -0,0 +1,188 @@
+			// either if the connection was supplied by user (let him handle that in such case),
+			// but there are currently no ways to know this from here.
+			if (!session.ConnectionManager.Transaction.IsActive)
+			{
+				// Enlist is called terribly frequently, and in some circumstances, it will
+				// not support to be called with the same value. So track what was the previous
+				// call and do not call it again if unneeded.
+				// (And Sql/OleDb/Odbc/Oracle manage/PostgreSql/MySql/Firebird/SQLite connections
+				// support multiple calls with the same ongoing transaction, but some others may not.)
+				var current = GetCurrentTransaction();
+				var connection = session.Connection;
+				System.Transactions.Transaction previous;
+				if (!_sessionsTransaction.TryGetValue(connection, out previous) || previous != current)
+				{
+					_sessionsTransaction.AddOrUpdate(connection, current, (s, t) => current);
+					if (current == null &&
+						// This will need an ad-hoc property on Dialect base class instead.
+						(session.Factory.Dialect is SQLiteDialect || session.Factory.Dialect is MsSqlCeDialect))
+					{
+						// Some connections does not support enlisting with null
+						// Let them with their previous transaction if any, the application
+						// will fail if the connection was left with a completed transaction due to this.
+						return;
+					}
+					session.Connection.EnlistTransaction(current);
+				}
+			}
+		}
+
+		public bool IsInDistributedActiveTransaction(ISessionImplementor session)
+		{
+			// Avoid agressive connection release while a transaction is ongoing. Allow
+			// auto-flushes (flushes before queries on dirtied entities).
+			return GetCurrentTransaction() != null;
+		}
+
+		public void ExecuteWorkInIsolation(ISessionImplementor session, IIsolatedWork work,
+			bool transacted)
+		{
+			using (var tx = new TransactionScope(TransactionScopeOption.Suppress))
+			{
+				// instead of duplicating the logic, we suppress the DTC transaction
+				// and create our own transaction instead
+				_adoNetTransactionFactory.ExecuteWorkInIsolation(session, work,
+					transacted);
+				tx.Complete();
+			}
+		}
+
+		private System.Transactions.Transaction GetCurrentTransaction()
+		{
+			try
+			{
+				return System.Transactions.Transaction.Current;
+			}
+			catch (InvalidOperationException)
+			{
+				// This damn thing may yield an invalid operation exception (instead of null
+				// or of a completed transaction) if we are between scope.Complete() and
+				// scope.Dispose(). This happen when having completed the scope then disposing
+				// the session and only after that disposing the scope.
+				// Instead of testing System.Transactions.Transaction.Current here, storing in
+				// connection manager the ambient transaction associated to the connection
+				// (and updating it when enlisting) then checking that stored transaction would
+				// reduce testes on Transaction.Current.
```

Comment:
Another bug fixed in custom transaction factory: accessing to Transaction.Current may throw, need to catch.